File: ical4j/src/main/java/org/calconnect/examples/SimpleEvent.java
/* ********************************************************************
    Appropriate copyright notice
*/
package org.calconnect.examples;

import net.fortuna.ical4j.model.Calendar;
import net.fortuna.ical4j.model.DateTime;
import net.fortuna.ical4j.model.PropertyList;
import net.fortuna.ical4j.model.TimeZone;
import net.fortuna.ical4j.model.component.VEvent;
import net.fortuna.ical4j.model.property.CalScale;
import net.fortuna.ical4j.model.property.Description;
import net.fortuna.ical4j.model.property.Duration;
import net.fortuna.ical4j.model.property.Location;
import net.fortuna.ical4j.model.property.ProdId;
import net.fortuna.ical4j.model.property.Version;

import java.util.GregorianCalendar;

/**
 * User: mike Date: 2/18/16 Time: 14:16
 */
public class SimpleEvent implements Example {
  @Override
  public String getKey() {
    return "simple-event";
  }

  @Override
  public String getTitle() {
    return "Simple 1 hour non-recurring event with no attendees";
  }

  @Override
  public Calendar getExample() {
    try {
      final TimeZone tz = Utils.getTimezone("America/New_York");
      
      /* Make up a start date 10 days from now.
         Ensure on an hour boundary
       */
      java.util.Calendar startDate = new GregorianCalendar();
      startDate.setTimeZone(tz);
      startDate.add(java.util.Calendar.DAY_OF_MONTH, 10);
      startDate.set(java.util.Calendar.MINUTE, 0);
      startDate.set(java.util.Calendar.SECOND, 0);

      // Create the event
      String eventName = "Simple event";
      DateTime start = new DateTime(startDate.getTime());
      start.setTimeZone(tz);
      
      VEvent event = new VEvent(start, eventName);

      PropertyList props = event.getProperties();

      /* One hour */
      Duration dur = new Duration(null, "PT1H");
      props.add(dur);

      /* Need a uid */
      props.add(Utils.generateUid());
      
      /* Add the location */
      props.add(new Location("Mo's bar - back room"));

      /* And the description - Note that ical4j handles any
         wrapping of long lines and with the 
         escaping of special characters. 
       */
      props.add(new Description(
              "We'll continue with the unfinished business from last time,\n" +
                      "in particular:\n" +
                      "Can names start wih a number?\n" +
                      "What if they are all numeric?\n" +
                      "Reuse of names - is it valid\n" +
                      "I remind the attendees we have spent 3 " +
                      "months on these subjects. We need " +
                      "closure!!!"));
      
      /* Create a calendar object */
      Calendar cal = new Calendar();
      PropertyList calProps = cal.getProperties();
      
      calProps.add(
              new ProdId("-//ABC Corporation/NONSGML iCal4j 1.0//EN"));
      calProps.add(CalScale.GREGORIAN);
      calProps.add(Version.VERSION_2_0);

      // Add the event and print
      cal.getComponents().add(event);
      
      return cal;
    } catch (Throwable t) {
      t.printStackTrace();
      return null;
    }
  }
}

File: libical/SimpleEvent.c
/* Example libical code to create Simple Event
 *
 * To compile: cc -o SimpleEvent -lical SimpleEvent.c
 */

#include <libical/ical.h>

#include <stdio.h>


int main(int argc, char *argv[])
{
    icalcomponent *cal, *event;
    icalproperty *prop;
    const char *my_product, *uid, *tzid, *desc;
    struct icaltimetype dtstart;
    struct icaldurationtype duration;

    /* Create VCALENDAR component */
    cal = icalcomponent_new(ICAL_VCALENDAR_COMPONENT);

    /* Create VERSION property and add to calendar.
       In this case, we create the new property and assign it to a variable
       (perhaps we need to access it later).
       We then pass this variable to the 'add' function.
    */
    prop = icalproperty_new_version("2.0");
    icalcomponent_add_property(cal, prop);

    /* Create PRODID property and add it to calendar.
       In this case, we don't plan on needing the property later
       so we can nest the 'new' function within the 'add' function.
     */
    my_product = "-//ABC Corporation/NONSGML My Product//EN";
    icalcomponent_add_property(cal, icalproperty_new_prodid(my_product));

    /* Create a uid for our event.
       NOTE: This should probably be created by a third-party library
       to guarantee uniqueness (e.g. OSSP UUID).
    */
    uid = "ff808181-1fd7389e-011f-d7389ef9-00000003";

    /* Create start time for our meeting */
    dtstart = icaltime_null_time();
    dtstart.year = 2016;
    dtstart.month = 4;
    dtstart.day = 20;
    dtstart.hour = 12;
    tzid = "America/New_York";

    /* Create duration for our meeting (1hr = 3600s) */
    duration = icaldurationtype_from_int(3600);

    /* Create VEVENT component.
       In this case, we use the varargs variants of creating a component
       and property which allows us to construct the component all in one go.
    */
    event =
        icalcomponent_vanew(ICAL_VEVENT_COMPONENT,
                            icalproperty_new_summary("Lunchtime meeting"),
                            icalproperty_new_uid(uid),
                            icalproperty_vanew_dtstart(dtstart,
                                                       icalparameter_new_tzid(tzid),
                                                       0),
                            icalproperty_new_duration(duration),
                            0);

    /* Create LOCATION property and add it to the event */
    icalcomponent_add_property(event,
                               icalproperty_new_location("Mo's bar - back room"));

    /* Create DESCRIPTION property and add it to the event */
    desc =
        "We'll continue with the unfinished business from last time,\n"
        "in particular:\n"
        "Can names start wih a number?\n"
        "What if they are all numeric?\n"
        "Reuse of names - is it valid\n"
        "I remind the attendees we have spent 3 "
        "months on these subjects. We need "
        "closure!!!";
    icalcomponent_add_property(event, icalproperty_new_description(desc));

    /* Add our event to the calendar */
    icalcomponent_add_component(cal, event);

    /* Create a string from our calendar and send it to stdout */
    printf("%s", icalcomponent_as_ical_string(cal));
}
